### DISCUSSION TITLE
ModelicaFormatMessage / ModelicaFormatError library for OM1.20.0 64-bit Windows

### DISCUSSION THREAD

**Discussion by matthewkindig** [OpenModelica/OpenModelica#10105] [Q&A] [answered]
I am integrating OpenModelica v.120.0 (64-bit Windows) with an existing C library through the ExternalObject interface. As I have several DLLs that are called within the C library, I am compiling my library into a DLL using Visual Studio (MSVC). This works as expected: I am able to successfully exchange data between the Modelica and C environments.

However, I see from the language specification ([[version 3.6-dev, section 12.9.6.1](https://specification.modelica.org/master/functions.html#external-function-interface)]) that I should be able to use the `ModelicaFormatMessage` and `ModelicaFormatError` functions to create formatted messages that will appear in my OM output file.  I have found the `ModelicaUtilities.h` header file containing those function signatures, however I have not found the corresponding binaries (DLL, .libs, etc.) that I need to link to in Visual Studio. Are these in a standard location within the OpenModelica install directory (OpenModelica1.20.0-64bit) ? I found "bin" and "lib" folders within that folder, but I'm not sure which of the files within there I should be linking to.  I've seen reference online to a `ModelicaExternalC` library, but I cannot find that anywhere. Any assistance you can provide would be greatly appreciated.

**Comment by AnHeuermann** [comment]
For my self-build (Ubuntu) OpenModelica there is `install/share/omc/sources/c/util/ModelicaUtilities.c` with the definitions of `ModelicaFormatMessage` and `ModelicaFormatError`.
For my Windows installation of OpenModelica the same file can be found in `C:\Program Files\OpenModelica1.21.0-dev-64bit\include\omc\c\util\ModelicaUtilities.c`. 

I checked (on my Ubunut build) with `nm -D` and found `ModelicaMessage` to be part of libOpenModelicaRuntimeC.so. I'm assuming on Windows (`C:\Program Files\OpenModelica1.21.0-dev-64bit\bin\libOpenModelicaRuntimeC.dll`) it is as well. You might experience some problems if libOpenModelicaRuntimeC.dll is compiled with gcc (from MINGW).

**Comment by casella** [comment, marked as answer]
This is a thorny issue, you can see an extensive discussion in  #9029 and modelica/ModelicaSpecification#2191. We had to solve this problem with the ExternalMedia library, that extensively relies on pre-compiled external binaries, and where it is pretty common that function calls are made to compute thermodynamic properties outside the range of validity of the model, which should throw a ModelicaError.

In principle, everything would be easy if both the Modelica tool and the simulation runtime exported symbols for ModelicaError and ModelicaWarning, so they could be imported and dynamically linked. Unfortunately this doesn't happen in all tools on all operating systems.

So, we came up with a workaround that is demonstrated in the source code of ExternalMedia, which works on Windows/Linux and on Dymola/OpenModelica (we haven't tried other Modelica tools yet). The idea is to explicitly pass pointers to the ModelicaError and ModelicaWarning functions to the external function implementations in the dynamic library.

In this specific case, the external function that gets called first, e.g., [setState_ph](https://github.com/modelica-3rdparty/ExternalMedia/blob/f524f9478d64cb5657acafd941056e43f52ad918/Modelica/ExternalMedia/Media/BaseClasses/ExternalTwoPhaseMedium.mo#L210), calls an external function `TwoPhaseMedium_setState_ph_C_impl_wrap` which is directly defined in the include annotation, so it's not in the DLL but it gets compiled by the Modelica tool when compiling the Modelica model, and therefore has static access to the ModelicaError and ModelicaWarning functions _as long as it is executed at runtime_, not at compile time.

The `TwoPhaseMedium_setState_ph_C_impl_wrap` function then calls the function [TwoPhaseMedium_setState_ph_C_impl_err](https://github.com/modelica-3rdparty/ExternalMedia/blob/f524f9478d64cb5657acafd941056e43f52ad918/Projects/Sources/externalmedialib.cpp#L547), which is instead compiled into the dynamic library, _explicitly passing to it the pointers to the ModelicaError and ModelicaWarning functions_. The TwoPhaseMedium_setState_ph_C_impl_err then [copies these pointers in two static pointers](https://github.com/modelica-3rdparty/ExternalMedia/blob/f524f9478d64cb5657acafd941056e43f52ad918/Projects/Sources/externalmedialib.cpp#L551) inside the dynamic library. Afterwards, all other functions that can be called after calling setState_ph can call the ModelicaError or ModelicaWarning function using these pointers, without the need of passing them the pointers all the time.

Note that this workaround is only required in Windows, since in Linux those symbols seem to be available both in Dymola and OpenModelica, so they can be directly imported and there is no need to explicitly pass function pointers around.

Also note that this workaround may not work if the external function is called at compile time, because in that case even `TwoPhaseMedium_setState_ph_C_impl_wrap` may not have access to the pointers to those two functions. This is not a problem in ExternalMedia, because some functions are indeed called at compile time (e.g., to compute package constants), but they never generate errors, while the functions that generate errors or warnings are always called at runtime during initialization or simulation.

Hope this helps 😃

**Reply by matthewkindig** [reply]
Thank you both @casella and @AnHeuermann for your insight. 

@AnHeuermann: I actually had added the /bin directory as a library directory in Visual Studio, and still ran into linker errors for these two functions. Since posting, I used [Dependencies.exe](https://github.com/lucasg/Dependencies) to export the names of all functions contained within DLLs in my OpenModelica install directory (recursively), and found that the ModelicaFormat* functions appear in a few different DLLs:  `libOpenModelicaRuntimeC.dll`, `libSimulationRuntimeC.dll`, `libOMCppModelicaUtilities.dll`, `OMCppModelicaUtilities.dll`. I have since added all of those directories as library directories to VS without success. My online searching indicates there is a way to extract *.lib files from a DLL for linking to VS, but I haven't fully investigated that approach yet.

@casella: It will take me a while to fully digest your answer, but it sounds like I should investigate the source code for `ExternalMedia` for some further insight into this issue. Once I've had a chance to fully understand how that code addresses this issue, I might post back.

Thanks again to you both for your help!

**Reply by casella** [reply]
@matthewkindig, it took me 15 years to come up with a solution for ExternalMedia that worked with two different tools on two different OS's, and I only managed when @fedetft started to delve into the issue - he really understands compilers and operating systems (unlike me, I'm a good modeller instead), he actually wrote a [fully Posix-compliant OS himself](https://github.com/fedetft/miosix-kernel) in his spare time 😄.

It took us several attempts and iterations to get this done properly, and it was definitely not trivial. I actually had to go through the ExternalMedia code again to recall what the final solution was. 

The bottom line is, to play safe it's best to pass the pointers to ModelicaError and ModelicaWarning explicitly to the functions compiled in the DLL. That is guaranteed to always work. In Linux, it seems that you can skip this because all required symbols are exported, but you cannot take that for granted in general, unless this requirement is stated in the Modelica Language Specification and adopted by all tool vendors on all OS's.

**Comment by mahge** [comment]
As @AnHeuermann mentioned you should link to `libOpenModelicaRuntimeC.dll`. This library is the one that contains the definitions for the function `ModelicaFormatMessage`.  This library should be located in the `bin/` folder in your OpenModelica installation.

If all you need is to call `ModelicaFormatMessage` from your code with **the assumption that OpenModelica is already available**, then this should be enough. If we do not run into other issues like `gcc` vs `MSVC` as @AnHeuermann mentioned. The case @casella mentioned above is a bit more complicated since it needed to be ambivalent to the Modelica compiler and the OS.

You can try that and let us know exactly what errors you encounter.
